Include return progress arg

burnout87 opened the pull request.

dsavchenko commented:
jobdir added to the response of nb2service for running jobs by https://github.com/oda-hub/nb2workflow/issues/129

burnout87 commented:
First round of review, mainly to understand if something is missing.

a small change in the dispatcher https://github.com/oda-hub/dispatcher-app/pull/632 will also be needed

dsavchenko commented on dispatcher_plugin_nb2workflow/products.py:
```diff
@@ -174,6 +174,27 @@ def write(self):
+
+class NB2WProgressProduct(NB2WProduct):
+    # TODO to be adapted
+    type_key = 'http://odahub.io/ontology#ODAProgressProduct'
```
`type_key` is not really needed as far as the instance is directly created, not with the factory

burnout87 replied:
and in which cases is it needed?
should it be just removed for this particular class ?

dsavchenko replied:
I don't see any reason to add such a term to the ontology, therefore there is no reason to keep it in the code.

If you just remove it, I think https://github.com/oda-hub/dispatcher-plugin-nb2workflow/blob/908cda39605dc3f62f3fc9773374b1c13cda47de/dispatcher_plugin_nb2workflow/products.py#L76 will break, but it can be easily adapted.

"Ordinary" products are all constructed according to their URIs by the factory function, which is called here https://github.com/oda-hub/dispatcher-plugin-nb2workflow/blob/908cda39605dc3f62f3fc9773374b1c13cda47de/dispatcher_plugin_nb2workflow/queries.py#L138, but this particular product is only initiated in the else clause

burnout87 replied:
true, `prod_list_factory` is not called for `NB2WProgressProduct`, so, at the moment, removal of `type_key` won't break

dsavchenko commented on dispatcher_plugin_nb2workflow/queries.py:
```diff
@@ -124,20 +125,28 @@ def get_data_server_query(self, instrument, config=None, **kwargs):
-        if 'output' in res.json().keys(): # in synchronous mode
-            _o_dict = res.json() 
+        res_content_type = res.headers.get('content-type', None)
+        if res_content_type is not None and res_content_type == 'application/json':
```
This works well, of course, but we can think about more explicit way to distinguish the *progress* response vs normal products

burnout87 replied:
I agree, it has to be improved. I had thought of adding an additional arg to the `build_product_list` but it would require to update such method for all the other plugins

dsavchenko replied:
Yes, something less intrusive would be better)

burnout87 replied:
by returning a dict containing the response, along with additional entries, in `get_progress_run`, the check can be done in a clean way

dsavchenko commented on dispatcher_plugin_nb2workflow/products.py:
```diff
@@ -174,6 +174,27 @@ def write(self):
+        self.name = name
+        self.progress_data = progress_html_data
+
+    def write(self):
+        file_path = os.path.join(self.out_dir, self.name)
+        with open(file_path, 'w') as fd:
+            fd.write(self.progress_data)
+        self.file_path = file_path
+
+    def get_html_draw(self):
+        # TODO not sure is correct, to verify
+        return {'image': {'div': '<br>' + self.progress_data, 'script': ''}}
```
This is not the standard product, so it will be treated by frontend in a special way, not embedded in the product view modal window. Probably, opening the html representation of the notebook in the new browser tab will be enough. If so, we can get rid of these methods here (will require an adaptation of `NB2WProductQuery.process_product_method()`)

burnout87 replied:
which type of adaption? in `process_product_method`, in case of a progressProduct product, those two methods are not used

dsavchenko replied:
It's treated here https://github.com/oda-hub/dispatcher-plugin-nb2workflow/blob/908cda39605dc3f62f3fc9773374b1c13cda47de/dispatcher_plugin_nb2workflow/queries.py#L164-L167 in case of oda_api request.

If it's frontend request, the processing is in this `else` clause https://github.com/oda-hub/dispatcher-plugin-nb2workflow/blob/908cda39605dc3f62f3fc9773374b1c13cda47de/dispatcher_plugin_nb2workflow/queries.py#L176
and there is no special treatment added there for this product.

dsavchenko replied:
But we can, of course, set `api=true` for this specific request even in frontend and not bother about this stuff

burnout87 replied:
I think it's better if I adapt the function accordingly, it will be more controllable from the side of the plugin

burnout87 replied:
`process_product_method` is now adapted, and `write` and `process_product_method` can actually be removed...

What do you think?

dsavchenko replied:
Yeah, let's remove them

dsavchenko commented on dispatcher_plugin_nb2workflow/dataserver_dispatcher.py:
```diff
@@ -98,6 +98,44 @@ def test_has_input_products(self, instrument, logger=None):
         query_out.set_done('input products check skipped')
         return query_out, []
 
+    def get_progress_run(self, **kwargs):
```
Why `**kwargs` and not an explicit list of arguments?

burnout87 replied:
I initially wanted to make it flexible with the args that will be passed from the dispatcher, but those will always be the same. 

So yes, no need for `kwargs`

dsavchenko commented on dispatcher_plugin_nb2workflow/dataserver_dispatcher.py:
```diff
@@ -98,6 +98,44 @@ def test_has_input_products(self, instrument, logger=None):
+        param_dict = kwargs.get('param_dict', self.param_dict)
+
+        run_asynch = kwargs.get('run_asynch', None)
+        call_back_url = kwargs.get('call_back_url', None)
```
Not sure if `_async_request_callback` arg is needed in this request

burnout87 replied:
let me know, I will remove it in case

dsavchenko replied:
It's not used to calculate the hash, by which the async job is identified. 
So if the job is already submitted to the async job queue by nb2service backend, the plugin will get the response with the status and jobdir regardless of this parameter. 

Some edge cases came to mind, I will summarise in the separate message

dsavchenko reviewed (changes requested):
That's nice overall. 

We need to agree on how this is treated by both frontend and oda_api to properly structure the dispatcher response, so we can get rid of unneded product methods (e.g. we don't really required to save this notebook representation into file with `write` if there won't be a "download" button)

dsavchenko commented:
Edge cases:
1. There is no such job in backend (normally, this doesn't happen, of course). In this case, we need to keep both  `_async_request_callback` and `_async_request` in the request, the job will be submitted just like with `run_query()`. May be, there is even no need for additional adaptations. But the jobdir will most probably be `None` (see case 2)
2. If the job is submitted to the async queue, but is not already running (status=submitted), the jobdir hasn't been set yet and will be returned as None. Needs separate treatment.
3. Synchronous request. Not sure if this will work at all (at least I didn't do any adaptations to nb2workflow regarding jobdir for synchronous case). Need to check. (But generally speaking, synchronous backend is only applicable to really short jobs, and there is no need to track progress. Probably it makes sense to just disable this method somehow in this case?)

burnout87 commented:
> Edge cases:
> 
>     1. There is no such job in backend (normally, this doesn't happen, of course). In this case, we need to keep both  `_async_request_callback` and `_async_request` in the request, the job will be submitted just like with `run_query()`. May be, there is even no need for additional adaptations. But the jobdir will most probably be `None` (see case 2)
> 
>     2. If the job is submitted to the async queue, but is not already running (status=submitted), the jobdir hasn't been set yet and will be returned as None. Needs separate treatment.
> 

So, if I understand correctl the `_async_request_callback` arg is needed.

>     3. Synchronous request. Not sure if this will work at all (at least I didn't do any adaptations to nb2workflow regarding jobdir for synchronous case). Need to check. (But generally speaking, synchronous backend is only applicable to really short jobs, and there is no need to track progress. Probably it makes sense to just disable this method somehow in this case?)

In the case of `run_asynch=False` (or `None`), `_async_request_callback` is not set. So, actually, in case of synch requests, it's working (I also added a test).

At the same time, I agree with you that a use-case where we check the progress of the execution for a sunch request is not really "realistic".

In any case, I added a test with the the synch case (`'run_asynch': False,`).

burnout87 commented:
>     2. If the job is submitted to the async queue, but is not already running (status=submitted), the jobdir hasn't been set yet and will be returned as None. Needs separate treatment.
> 

Actually, I already consider this scenario: `if status == 'started' or status == 'done'` then I extract the `job_dir`.

Otherwise, `status=submitted`, the result of the call will be empty (`res_trace=None`)

dsavchenko commented on dispatcher_plugin_nb2workflow/queries.py:
```diff
@@ -160,26 +177,34 @@ def process_product_method(self, instrument, prod_list, api=False):
-                im = product.get_html_draw()
-                if im:
-                    image_list.append(im)
+                html_draw = product.progress_data
```
Something wrong here, this should be for ProgressProduct only

And now there is no call for `get_html_draw()` of other products

burnout87 replied:
yes, I am fixing it now

dsavchenko replied:
Hm, but tests have passed. Proper html image generation for normal products is not covered. Added an issue https://github.com/oda-hub/dispatcher-plugin-nb2workflow/issues/77

dsavchenko commented:
> So, if I understand correctl the `_async_request_callback` arg is needed.

Seems so, finally

dsavchenko commented on tests/test_plugin.py:
```diff
@@ -489,4 +489,62 @@ def test_failed_nbhtml_download(live_nb2service,
+
+
+@pytest.mark.parametrize("api", [True, False])
+def test_return_progress(dispatcher_live_fixture, mock_backend, api):
```
duplicate test name

dsavchenko reviewed (changes requested):
Only one fix and cool overall!

I only wonder how the response will then be used on the frontend side. We have full page html in json, how to show it there?

volodymyrss commented on dispatcher_plugin_nb2workflow/products.py:
```diff
@@ -32,6 +32,7 @@ def write(self, file_name=None, overwrite=True, file_dir=None):
         self.table_data.write(file_path, overwrite=overwrite, format='ascii.ecsv')
         
 class NB2WProduct:  # TODO: decide on the name precedence 
+    progress_product = False
```
```suggestion
    progress_product = False
    
```

burnout87 replied:
this actually has to be removed

volodymyrss commented on dispatcher_plugin_nb2workflow/products.py:
```diff
@@ -72,25 +73,25 @@ def _init_as_list(cls, encoded_data, *args, **kwargs):
+                                                                           name=key,
                                                                            **extra_kw
-                                                                           ) 
+                                                                           )
```
a bunch of changes here just add spaces

burnout87 replied:
probably introduced by the IDEs during some copy/paste operations

volodymyrss replied:
this code was already there, strange that you needed to copy-paste it. But anyway, could remove the copy-paste artifacts.

volodymyrss commented on dispatcher_plugin_nb2workflow/dataserver_dispatcher.py:
```diff
@@ -98,6 +98,52 @@ def test_has_input_products(self, instrument, logger=None):
+            task = self.task
+        if param_dict is None:
+            param_dict = self.param_dict
+        if run_asynch is not None and run_asynch and call_back_url is not None:
```
wouldn't this be the same

```suggestion
        if run_asynch and call_back_url is not None:
```

volodymyrss commented on dispatcher_plugin_nb2workflow/dataserver_dispatcher.py:
```diff
@@ -98,6 +98,52 @@ def test_has_input_products(self, instrument, logger=None):
+                jobdir = resroot['jobdir'].split('/')[-1]
+                trace_url = os.path.join(self.data_server_url, 'trace', jobdir, task.strip('/'))
+                res_trace = requests.get(trace_url)
+                if res_trace_dict is None:
```
how can it be not `None`?

volodymyrss replied:
yes, so it's always `None` when you get here, why condition?

volodymyrss commented on dispatcher_plugin_nb2workflow/dataserver_dispatcher.py:
```diff
@@ -98,6 +98,52 @@ def test_has_input_products(self, instrument, logger=None):
+                query_out.set_failed('Error in the backend',
+                                     message='connection status code: ' + str(res.status_code),
+                                     extra_message=res.json()['exceptions'][0])
+            except:
```
please capture Exception, report in sentry, in log

burnout87 replied:
sentry is not used within this plugin...or at least I haven't seen it

burnout87 replied:
in any case, a sentry is sent from the `set_query_exception` method from the dispatcher, with the details of the exception

volodymyrss replied:
through failed state of query? Then please still catch the exception and provide it to the structure which will be eventually sent to sentry. Right now you are just ignoring some exceptions, `except:` is basically never a good thing.

burnout87 replied:
that except was there in case the response from the backend was not json-formatted, I will re-write it to make it cleaner

burnout87 replied:
it's better now....

burnout87 replied:
> through failed state of query? 

yes, here: https://github.com/oda-hub/dispatcher-app/blob/7257b06ebe157878091ad139eb59361acb65c467/cdci_data_analysis/analysis/products.py#L165-L220

volodymyrss commented on dispatcher_plugin_nb2workflow/queries.py:
```diff
@@ -124,20 +125,33 @@ def get_data_server_query(self, instrument, config=None, **kwargs):
+            res_progress_product = res.get('progress_product', False)
+            res = res.get('res', None)
+        res_content_type = res.headers.get('content-type', None)
+        if res_content_type is not None and res_content_type == 'application/json':
```
can you make a comment about what does this condition actually represent
